policy-controller/k8s/index/src/index.rs

```diff
@@ -0,0 +1,738 @@
+    /// a default is server is created.
+    pub fn pod_server_rx(
+        &mut self,
+        namespace: &str,
+        pod: &str,
+        port: u16,
+    ) -> Result<watch::Receiver<InboundServer>> {
+        let ns = self
+            .namespaces
+            .by_ns
+            .get_mut(namespace)
+            .ok_or_else(|| anyhow::anyhow!("namespace not found: {}", namespace))?;
+        let pod = ns
+            .pods
+            .by_name
+            .get_mut(pod)
+            .ok_or_else(|| anyhow::anyhow!("pod {}.{} not found", pod, namespace))?;
+        Ok(pod
+            .port_server_or_default(port, &self.cluster_info)
+            .rx
+            .clone())
+    }
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::Pod> for Index {
+    #[instrument(skip_all, name = "pod.apply", fields(
+        ns = %pod.namespace().unwrap(),
+        pod = %pod.name(),
+    ))]
+    fn apply(&mut self, pod: k8s::Pod) {
+        let namespace = pod.namespace().unwrap();
+        let name = pod.name();
+        let port_names = pod::tcp_port_names(pod.spec);
+        let meta = pod::Meta::from_metadata(pod.metadata);
+
+        // Add or update the pod. If the pod was not already present in the
+        // index with the same metadata, index it against the policy resources,
+        // updating its watches.
+        let ns = self.namespaces.get_or_default(namespace);
+        match ns.pods.update(name, meta, port_names) {
+            Ok(None) => {}
+            Ok(Some(pod)) => pod.reindex_servers(&ns.policy),
+            Err(error) => {
+                tracing::error!(%error, "Illegal pod update");
+            }
+        }
+    }
+
+    #[instrument(skip(self), name = "srv.delete")]
+    fn delete(&mut self, namespace: String, pod: String) {
+        if let Entry::Occupied(mut ns) = self.namespaces.by_ns.entry(namespace) {
+            // Once the pod is removed, there's nothing else to update. Any open
+            // watches will complete.  No other parts of the index need to be
+            // updated.
+            if ns.get_mut().pods.by_name.remove(&pod).is_some() && ns.get().is_empty() {
+                ns.remove();
+            }
+        }
+    }
+
+    // Since apply only reindexes a single pod at a time, there's no need to
+    // handle resets specially.
+}
+
+impl kubert::index::IndexNamespacedResource<k8s::policy::Server> for Index {
+    #[instrument(skip_all, name = "srv.apply", fields(
+        ns = %srv.namespace().unwrap(),
+        srv = %srv.name(),
+    ))]
+    fn apply(&mut self, srv: k8s::policy::Server) {
+        let namespace = srv.namespace().expect("server must be namespaced");
+        let name = srv.name();
+        let server = server::Server::from_resource(srv, &self.cluster_info);
+        self.namespaces
+            .get_or_default_with_reindex(namespace, |ns| ns.policy.update_server(name, server))
+    }
+
+    #[instrument(skip(self), name = "srv.delete")]
```

Comment:
nit/tioli: would we want to include the ns and service name here?
```suggestion
    #[instrument(skip(self), name = "srv.delete", fields(ns = %ns, srv = %srv)]
```